tap_mailchimp/streams.py

```diff
@@ -0,0 +1,538 @@
+        'sort_dir': 'ASC'
+    }
+    child = ['reports_email_activity', 'unsubscribes']
+    report_streams = ['reports_email_activity']
+    data_key = stream_name
+
+
+class Unsubscribes(FullTable):
+    """Class for 'unsubscribes' stream"""
+    stream_name = 'unsubscribes'
+    key_properties = ['campaign_id', 'email_id']
+    path = '/reports/{}/unsubscribed'
+    streams_to_sync = ['campaigns']
+    data_key = stream_name
+
+
+class ReportEmailActivity(Incremental):
+    """Class for 'reports_email_activity' stream"""
+    stream_name = 'reports_email_activity'
+    extra_fields = ['emails.activity']
+    key_properties = ['campaign_id', 'action', 'email_id', 'timestamp']
+    streams_to_sync = ['campaigns']
+    path = '/reports/{}/email-activity'
+    data_key = 'emails'
+    replication_keys = ['timestamp']
+
+    def write_activity_batch_bookmark(self, batch_id):
+        """Write batch id bookmark"""
+        self.write_bookmark(['reports_email_activity_last_run_id'], batch_id)
+
+    def write_email_activity_chunk_bookmark(self, current_bookmark, current_index, sorted_campaigns):
+        """Write chunk bookmark for email activities"""
+        # Bookmark the next chunk because the current chunk will be saved in batch_id
+        # Index is relative to current bookmark
+        next_chunk = current_bookmark + current_index + 1
+        if next_chunk * EMAIL_ACTIVITY_BATCH_SIZE < len(sorted_campaigns):
+            self.write_bookmark(['reports_email_activity_next_chunk'], next_chunk)
+        else:
+            self.write_bookmark(['reports_email_activity_next_chunk'], 0)
+
+    def get_batch_info(self, batch_id):
+        """Function to get batch status"""
+        try:
+            return self.client.get('/batches/{}'.format(batch_id), endpoint='get_batch_info')
+        except HTTPError as e:
+            if e.response.status_code == 404:
+                raise BatchExpiredError('Batch {} expired'.format(batch_id))
+            raise e
+
+    def check_and_resume_email_activity_batch(self):
+        """Function to resume batch syncing from previous sync"""
+        batch_id = self.get_bookmark(['reports_email_activity_last_run_id'], None)
+
+        if batch_id:
+            try:
+                data = self.get_batch_info(batch_id)
+                if data['status'] == 'finished' and not data['response_body_url']:
+                    LOGGER.info('reports_email_activity - Previous run from state (%s) is empty, retrying.',
+                                batch_id)
+                    return
+            except BatchExpiredError:
+                LOGGER.info('reports_email_activity - Previous run from state expired: %s',
+                            batch_id)
+                return
+
+            # Resume from bookmarked job_id, then if completed, issue a new batch for processing.
+            campaigns = []  # Don't need a list of campaigns if resuming
+            self.sync_email_activities(campaigns, batch_id)
```

Comment:
Can we move these 2 lines to try and remove the return line?

Reply:
Also add unit test for the same

Reply:
Added unittest.